src/qiboml/operations/differentiation.py

```diff
@@ -152,7 +67,109 @@ def circuit(nqubits = 1):
+            circuits are executed.
+        initial_state (ndarray, optional): initial state on which the circuit
+            acts. Default is ``None``.
+        nshots (int, optional): number of shots if derivative is evaluated on
+            hardware. If ``None``, the simulation mode is executed.
+            Default is ``None``.
+
+    Returns:
+        (float): Gradient of the expectation value of the hamiltonian
+            with respect to circuit's variational parameters.
+
+    """
+    return [
+        _one_parameter_shift(
+            hamiltonian=hamiltonian,
+            circuit=circuit,
+            parameter_index=i,
+            initial_state=initial_state,
+            nshots=nshots,
+            exec_backend=exec_backend,
+        )
+        for i in range(len(circuit.get_parameters()))
+    ]
+
+
+def symbolical(
+    hamiltonian: qibo.hamiltonians.Hamiltonian,
+    circuit: qibo.Circuit,
+    exec_backend: qibo.backends.Backend,
+    initial_state: Optional[Union[np.ndarray, qibo.Circuit]] = None,
+):
+    """
+    Compute the gradient of the expectation value of ``hamiltonian`` over the
+    state we get executing ``circuit`` over ``initial_state`` with a certain
+    ``nshots`` w.r.t. the parameters of the circuit using TensorFlow automatic
+    differentiation.
+
+    Args:
+        circuit (:class:`qibo.models.circuit.Circuit`): custom quantum circuit.
+        hamiltonian (:class:`qibo.hamiltonians.Hamiltonian`): target observable.
+            if you want to execute on hardware, a symbolic hamiltonian must be
+            provided as follows (example with Pauli Z and ``nqubits=1``):
+            ``SymbolicHamiltonian(np.prod([ Z(i) for i in range(1) ]))``.
+        execution_backend (qibo.backends.Backend): Qibo backend on which the
+            circuits are executed.
+        initial_state (ndarray, optional): initial state on which the circuit
+            acts. Default is ``None``.
+
+    Returns:
+        (float): Gradient of the expectation value of the hamiltonian
+            with respect to circuit's variational parameters.
+
+    """
+    import tensorflow as tf  # pylint: disable=import-error
+
+    # TODO: how to fix this at lower level?
+    circuit_parameters = tf.Variable(circuit.get_parameters(), dtype="float64")
+
+    with tf.GradientTape() as tape:
+        circuit.set_parameters(circuit_parameters)
+        expval = expectation._exact(hamiltonian, circuit, initial_state, exec_backend)
+
+    gradients = tape.gradient(expval, circuit_parameters)
+
+    return tf.squeeze(gradients)
+
+
+def symbolical_with_jax(
+    hamiltonian: qibo.hamiltonians.Hamiltonian,
+    circuit: qibo.Circuit,
+    exec_backend: qibo.backends.Backend,
+    initial_state: Optional[Union[np.ndarray, qibo.Circuit]] = None,
+):
+    def _expectation(params):
+        params = jax.numpy.array(params)
+        circuit.set_parameters(params)
+        return expectation._exact(hamiltonian, circuit, initial_state, exec_backend)
```

Comment:
This is the function I was telling you yesterday: if this was already available, you didn't have to write it here, but just pass this function as input, and then this whole function would be just `jax.grad()`, plus the manipulations you need to make it compatible.

I.e. all the inputs of `symbolical_with_jax` are used in a single place, so the only input you need is the result of the operation in that place.